lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -1014,6 +1015,129 @@ class RAIIOpenedArchetypesTracker {
   }
 };
 
+/// Helper routine for propagating concrete type from
+/// class hierarchy analysis.
+bool SILCombiner::propagateConcreteTypeFromCTAInternal(
+    FullApplySite Apply, ProtocolDecl *Protocol, SILValue &Arg,
+    SILValue &NewArg, Optional<ProtocolConformanceRef> &ConformanceRef,
+    ArchetypeType *OpenedArchetype, CanType &ConcreteType) {
+
+  if (VisitedAIForCTA.find(Apply) != VisitedAIForCTA.end())
+    return false;
+  auto &M = Builder.getModule();
+  if (!Protocol)
+    return false;
+  auto *NTD = CTA->getSoleTypeImplementingProtocol(Protocol);
+  if (!NTD)
+    return false;
+
+  ClassDecl *CD = nullptr;
+  /// Check if the class has no subclasses: direct or indirect.
+  if ((CD = dyn_cast<ClassDecl>(NTD)) &&
+          ((CHA->hasKnownDirectSubclasses(CD)) ||
+          (CHA->hasKnownIndirectSubclasses(CD)))) {
+    return false;
+  }
+
+  auto ElementType = NTD->getDeclaredType()->getCanonicalType();
+  if ((isa<ClassDecl>(NTD)) && (!(ElementType->getClassOrBoundGenericClass()))) {
+    return false;
+  }
+
+  bool isCopied = false;
+  if (auto *Instance = dyn_cast<AllocStackInst>(Arg)) {
+    if (SILValue Src =
+            getAddressOfStackInit(Instance, Apply.getInstruction(), isCopied)) {
+      Arg = Src;
+    }
+  }
+  if (auto *OER = dyn_cast<OpenExistentialRefInst>(Arg)) {
+    ConcreteType = ElementType;
+    SILType ConcreteSILType = M.Types.getLoweredType(ConcreteType);
+    auto *URCI =
+        Builder.createUncheckedRefCast(OER->getLoc(), OER, ConcreteSILType);
+    ConformanceRef = M.getSwiftModule()->lookupConformance(
+        ConcreteType, Protocol);
+    OpenedArchetype =
+        OER->getType().getASTType()->castTo<ArchetypeType>();
+    NewArg = URCI;
+  } else if (auto *OEA = dyn_cast<OpenExistentialAddrInst>(Arg)) {
+    ConcreteType = ElementType;
+    SILType ConcreteSILType = SILType::getPrimitiveAddressType(ConcreteType);
```

Comment:
This will produce an invalid SIL type if its an Optional. You need to lower the SIL type; SILModule.Types.getLoweredType(ConcreteType)